### PR TITLE
Set label destination_service_name and destination_service_namespace to L4 metrics

### PR DESCRIPTION
**Pull request by imroc**
Fixes [#46169](https://github.com/istio/istio/issues/46169)

With this PR, we can see `Service Nodes` graph in kiali.

### CODE REVIEW

**Review by EItanya — COMMENTED** [review]
makes sense to me

File: src/proxy/outbound.rs [hunk 1]
```diff
@@ -165,8 +165,8 @@ impl OutboundConnection {
                 metrics::SecurityPolicy::unknown
             },
             destination_service: None,
```
Comment (howardjohn) [inline src/proxy/outbound.rs:167]: I think we want this one as well?

File: src/proxy/outbound.rs [hunk 2]
```diff
@@ -512,6 +520,8 @@ struct Request {
     // The intended destination workload. This is always the original intended target, even in the case
     // of other proxies along the path.
     destination_workload: Option<Workload>,
+    destination_service_name: Option<String>,
```
Comment (howardjohn) [inline src/proxy/outbound.rs:523]: can we make this a Service object with name, namespace, and hostname? then we just have one

`destination_service: Option<Service>`

Comment (howardjohn) [reply src/proxy/outbound.rs:523]: Actually we don't need to make one, we already have Service in the `state` package. Unless it is hard to move around due to ownership - if so, then just make something like ServiceReference or whatever with those 3 fields

**Review by howardjohn — COMMENTED** [review]
lgtm, just need to set destination_service and refactor a bit. Thanks!

File: src/proxy/outbound.rs (hunk 1 above)
Comment (imroc) [reply src/proxy/outbound.rs:167]: I think `destination_service` is the hostname in L7,  In ztunnel, we can't get the hostname info of L7?

Comment (howardjohn) [reply src/proxy/outbound.rs:167]: We can - its already in the Service object.

We get a service VIP, and lookup the `Service` from the VIP. This has name, hostname, namespace.

Its not the `Host` header of the request we are reporting (well, not directly -- often it is the same)

Comment (imroc) [reply src/proxy/outbound.rs]: I see. Changed to use `Arc` to reference `Service`, and set `destination_service`, `destination_service_name`, `destination_service_namespece` with the `hostname`, `name` and `namespace` of `Service`.

File: src/state.rs [hunk 3]
```diff
@@ -41,11 +41,12 @@ pub mod policy;
 pub mod service;
 pub mod workload;
 
-#[derive(Debug, Hash, Eq, PartialEq, Clone, serde::Serialize)]
```
Comment (imroc) [inline src/state.rs:44]: The Hash trait cannot been derived for `Option<Arc<Service>>`, and I think this trait is not been used by Upstream, so I removed it. @howardjohn

**Comment by imroc** [comment]
Now it's refactored to use `Arc` to refer the `Service`, and set `destination_service`, `destination_service_name`, `destination_service_namespece` with the `hostname`, `name` and `namespace` of `Service`.

I've tested in my environment, it works perfectly, here is some of the metrics for productpage pod:

```txt
istio_tcp_connections_opened_total{reporter="source",source_workload="productpage-v1",source_canonical_service="productpage",source_canonical_revision="v1",source_workload_namespace="mesh-test",source_principal="spiffe://cluster.local/ns/mesh-test/sa/bookinfo-productpage",source_app="productpage",source_version="v1",source_cluster="cls-1bnt4qt5",destination_service="reviews.mesh-test.svc.cluster.local",destination_service_namespace="mesh-test",destination_service_name="reviews",destination_workload="reviews-v3",destination_canonical_service="reviews",destination_canonical_revision="v3",destination_workload_namespace="mesh-test",destination_principal="spiffe://cluster.local/ns/mesh-test/sa/bookinfo-reviews",destination_app="reviews",destination_version="v3",destination_cluster="cls-1bnt4qt5",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls"} 2887
istio_tcp_connections_closed_total{reporter="source",source_workload="productpage-v1",source_canonical_service="productpage",source_canonical_revision="v1",source_workload_namespace="mesh-test",source_principal="spiffe://cluster.local/ns/mesh-test/sa/bookinfo-productpage",source_app="productpage",source_version="v1",source_cluster="cls-1bnt4qt5",destination_service="reviews.mesh-test.svc.cluster.local",destination_service_namespace="mesh-test",destination_service_name="reviews",destination_workload="reviews-v3",destination_canonical_service="reviews",destination_canonical_revision="v3",destination_workload_namespace="mesh-test",destination_principal="spiffe://cluster.local/ns/mesh-test/sa/bookinfo-reviews",destination_app="reviews",destination_version="v3",destination_cluster="cls-1bnt4qt5",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls"} 2887
istio_tcp_received_bytes_total{reporter="source",source_workload="productpage-v1",source_canonical_service="productpage",source_canonical_revision="v1",source_workload_namespace="mesh-test",source_principal="spiffe://cluster.local/ns/mesh-test/sa/bookinfo-productpage",source_app="productpage",source_version="v1",source_cluster="cls-1bnt4qt5",destination_service="reviews.mesh-test.svc.cluster.local",destination_service_namespace="mesh-test",destination_service_name="reviews",destination_workload="reviews-v3",destination_canonical_service="reviews",destination_canonical_revision="v3",destination_workload_namespace="mesh-test",destination_principal="spiffe://cluster.local/ns/mesh-test/sa/bookinfo-reviews",destination_app="reviews",destination_version="v3",destination_cluster="cls-1bnt4qt5",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls"} 632255
istio_tcp_sent_bytes_total{reporter="source",source_workload="productpage-v1",source_canonical_service="productpage",source_canonical_revision="v1",source_workload_namespace="mesh-test",source_principal="spiffe://cluster.local/ns/mesh-test/sa/bookinfo-productpage",source_app="productpage",source_version="v1",source_cluster="cls-1bnt4qt5",destination_service="reviews.mesh-test.svc.cluster.local",destination_service_namespace="mesh-test",destination_service_name="reviews",destination_workload="reviews-v3",destination_canonical_service="reviews",destination_canonical_revision="v3",destination_workload_namespace="mesh-test",destination_principal="spiffe://cluster.local/ns/mesh-test/sa/bookinfo-reviews",destination_app="reviews",destination_version="v3",destination_cluster="cls-1bnt4qt5",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls"} 1729313
```

And our self-built UI also displayed service topology graph by service dimension perfectly without sidecar (only ztunnel, and use kiali to compute the topology)

image

@howardjohn PTAL

**Review by howardjohn — APPROVED** [review]

**Pull request merged**